compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/ClassReference.kt

```diff
@@ -0,0 +1,162 @@
+package com.squareup.anvil.compiler.internal
+
+import com.squareup.anvil.annotations.ExperimentalAnvilApi
+import com.squareup.anvil.compiler.api.AnvilCompilationException
+import com.squareup.anvil.compiler.internal.ClassReference.Descriptor
+import com.squareup.anvil.compiler.internal.ClassReference.Psi
+import org.jetbrains.kotlin.descriptors.ClassDescriptor
+import org.jetbrains.kotlin.descriptors.ModuleDescriptor
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.KtClassOrObject
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+
+/**
+ * Used to create a common type between [KtClassOrObject] class references and [ClassDescriptor]
+ * references, to streamline parsing.
+ *
+ * @see allSuperTypeClassReferences
+ * @see requireClassReference
+ */
+@ExperimentalAnvilApi
+public sealed class ClassReference {
+
+  public abstract val fqName: FqName
+
+  public class Psi(
+    public val clazz: KtClassOrObject,
+    override val fqName: FqName
+  ) : ClassReference()
+
+  public class Descriptor(
+    public val clazz: ClassDescriptor,
+    override val fqName: FqName
+  ) : ClassReference()
+
+  override fun toString(): String {
+    return "${this::class.qualifiedName}($fqName)"
+  }
+
+  override fun equals(other: Any?): Boolean {
+    if (this === other) return true
+    if (other !is ClassReference) return false
+
+    if (fqName != other.fqName) return false
+
+    return true
+  }
+
+  override fun hashCode(): Int {
+    return fqName.hashCode()
+  }
+}
+
+/**
+ * Attempts to resolve the [fqName] to a [ClassDescriptor] first, then falls back to a
+ * [KtClassOrObject] if the descriptor resolution fails. This will happen if the code being parsed
+ * was generated as part of the compilation round for this module.
+ */
+@ExperimentalAnvilApi
+public fun ModuleDescriptor.classReferenceOrNull(fqName: FqName): ClassReference? {
+  return fqName.classDescriptorOrNull(this)
+    ?.let { Descriptor(it, fqName) }
+    ?: getKtClassOrObjectOrNull(fqName)
+      ?.let { Psi(it, fqName) }
+}
+
+@ExperimentalAnvilApi
+public fun ClassDescriptor.toClassReference(): Descriptor {
+  return ClassReference.Descriptor(this, fqNameSafe)
+}
+
+@ExperimentalAnvilApi
+public fun KtClassOrObject.toClassReference(): Psi {
+  return ClassReference.Psi(this, requireFqName())
```

Comment:
The IDE show me here and above that `ClassReference.` is a redundant qualifier name and can be dropped.